python/ffsim/tenpy/hamiltonians/molecular_hamiltonian.py

```diff
@@ -0,0 +1,142 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+from __future__ import annotations
+
+import itertools
+
+import numpy as np
+from tenpy.models.lattice import Lattice
+from tenpy.models.model import CouplingMPOModel
+from tenpy.networks.site import SpinHalfFermionSite
+
+from ffsim.hamiltonians.molecular_hamiltonian import MolecularHamiltonian
+
+# ignore lowercase variable checks to maintain TeNPy naming conventions
+# ruff: noqa: N806
+
+
+class MolecularHamiltonianMPOModel(CouplingMPOModel):
+    """Molecular Hamiltonian."""
+
+    def init_sites(self, params):
+        cons_N = params.get("cons_N", "N", expect_type=str)
+        cons_Sz = params.get("cons_Sz", "Sz", expect_type=str)
+        site = SpinHalfFermionSite(cons_N=cons_N, cons_Sz=cons_Sz)
+        return site
+
+    def init_lattice(self, params):
+        L = params.get("L", 1, expect_type=int)
+        norb = params.get("norb", None, expect_type=int)
+        site = self.init_sites(params)
+        basis = np.array(([norb, 0.0], [0, 1]))
+        pos = np.array([[i, 0] for i in range(norb)])
+        lat = Lattice(
+            [L, 1],
+            [site] * norb,
+            order="default",
+            bc="open",
+            bc_MPS="finite",
+            basis=basis,
+            positions=pos,
+        )
+        return lat
+
+    def init_terms(self, params):
+        dx0 = np.array([0, 0])
+        norb = params.get("norb", None, expect_type=int)
+        one_body_tensor = params.get(
+            "one_body_tensor", np.zeros((norb, norb)), expect_type="array"
+        )
+        two_body_tensor = params.get(
+            "two_body_tensor", np.zeros((norb, norb, norb, norb)), expect_type="array"
+        )
+        constant = params.get("constant", 0, expect_type="real")
+
+        for p, q in itertools.product(range(norb), repeat=2):
+            h1 = one_body_tensor[q, p]
+            if p == q:
+                self.add_onsite(h1, p, "Nu")
+                self.add_onsite(h1, p, "Nd")
+                self.add_onsite(constant / norb, p, "Id")
+            else:
+                self.add_coupling(h1, p, "Cdu", q, "Cu", dx0)
+                self.add_coupling(h1, p, "Cdd", q, "Cd", dx0)
+
+            for r, s in itertools.product(range(norb), repeat=2):
+                h2 = two_body_tensor[q, p, s, r]
+                if p == q == r == s:
+                    self.add_onsite(0.5 * h2, p, "Nu")
+                    self.add_onsite(-0.5 * h2, p, "Nu Nu")
+                    self.add_onsite(0.5 * h2, p, "Nu")
+                    self.add_onsite(-0.5 * h2, p, "Cdu Cd Cdd Cu")
+                    self.add_onsite(0.5 * h2, p, "Nd")
+                    self.add_onsite(-0.5 * h2, p, "Cdd Cu Cdu Cd")
+                    self.add_onsite(0.5 * h2, p, "Nd")
+                    self.add_onsite(-0.5 * h2, p, "Nd Nd")
+                else:
+                    self.add_multi_coupling(
```

Comment:
I feel that this looping logic can be made a bit clearer by performing the
```
if p == q:
```
and
```
if p == q == r == s:
```
blocks in a separate loop, preceding the main loop. I think doing the one- and two-body tensors in separate loops will also improve the readability.

Reply:
Agree that this would improve readability. However, wouldn't both of these suggestions introduce more loops and therefore make the code slower?

Reply:
The total number of items being looped over is the same, so I expect the difference to be negligible. Besides, one of the reasons I ask for this change is that I think it will make it simpler to edit the logic later to reduce redundancy in the loop by taking account of symmetry. For example, since the one-body tensor is Hermitian, you only need to loop over its upper triangle. The two-body tensor also satisfies symmetries as demonstrated in [this test](https://github.com/qiskit-community/ffsim/blob/e509eeca3c1e89e51977d5792f4dbd780299fcda/tests/python/random_test.py#L77-L85)

Reply:
I have split up the loops, but did not manage to finish simplifying the `two_body_tensor` loop part yet.